File: mobile/lib/domain/community/place_on_track.dart
/// Track für POST /api/community/places: inkl. letztem Punkt, max. 40 Samples.
List<List<double>> parseTrackSamples(Object? raw, {int max = 40}) {
  if (raw is! List) return const [];
  final all = <List<double>>[];
  for (final p in raw) {
    if (p is! List || p.length < 2) continue;
    final lng = (p[0] as num?)?.toDouble();
    final lat = (p[1] as num?)?.toDouble();
    if (lng == null || lat == null) continue;
    if (!lng.isFinite || !lat.isFinite) continue;
    if (lat.abs() > 90 || lng.abs() > 180) continue;
    all.add([lng, lat]);
  }
  return sampleTrackLngLat(all, max: max);
}

List<List<double>> sampleTrackLngLat(
  List<List<double>> coords, {
  int max = 40,
}) {
  if (coords.length <= max) return List<List<double>>.from(coords);
  final out = <List<double>>[];
  for (var i = 0; i < max; i++) {
    final idx = i == max - 1
        ? coords.length - 1
        : ((i * (coords.length - 1)) / (max - 1)).round();
    final pt = coords[idx];
    if (out.isNotEmpty && out.last[0] == pt[0] && out.last[1] == pt[1]) {
      continue;
    }
    out.add(pt);
  }
  return out;
}

List<List<double>> trackMapsToLngLat(List<Map<String, dynamic>> track) {
  return [
    for (final p in track)
      if (p['lat'] is num && p['lng'] is num)
        [
          (p['lng'] as num).toDouble(),
          (p['lat'] as num).toDouble(),
        ],
  ];
}

File: src/lib/community/placeOnTrack.ts
/**
 * User-Ort nur mit Ride-Beweis: Punkt muss auf der mitgeschickten Spur liegen.
 */

import { projectOntoRoute, pointAlongRoute } from "@/lib/routing/routeProgress";

export const PLACE_SNAP_MAX_M = 80;

export type TrackLngLat = [number, number];

export function parseTrackSamples(raw: unknown, max = 40): TrackLngLat[] {
  if (!Array.isArray(raw)) return [];
  const all: TrackLngLat[] = [];
  for (const p of raw) {
    if (!Array.isArray(p) || p.length < 2) continue;
    const lng = Number(p[0]);
    const lat = Number(p[1]);
    if (!Number.isFinite(lat) || !Number.isFinite(lng)) continue;
    if (Math.abs(lat) > 90 || Math.abs(lng) > 180) continue;
    all.push([lng, lat]);
  }
  if (all.length <= max) return all;
  const out: TrackLngLat[] = [];
  for (let i = 0; i < max; i++) {
    const idx =
      i === max - 1
        ? all.length - 1
        : Math.round((i * (all.length - 1)) / (max - 1));
    const pt = all[idx];
    const last = out[out.length - 1];
    if (last && last[0] === pt[0] && last[1] === pt[1]) continue;
    out.push(pt);
  }
  return out;
}

export function snapPlaceOntoTrack(
  lat: number,
  lng: number,
  track: TrackLngLat[],
  maxOffM = PLACE_SNAP_MAX_M
): { lat: number; lng: number; alongM: number; offM: number } | null {
  if (track.length < 2) return null;
  if (!Number.isFinite(lat) || !Number.isFinite(lng)) return null;
  const p = projectOntoRoute(track, lat, lng);
  if (p.crossTrackM > maxOffM) return null;
  const pt = pointAlongRoute(track, p.distanceAlongM);
  return {
    lng: pt[0],
    lat: pt[1],
    alongM: p.distanceAlongM,
    offM: p.crossTrackM,
  };
}
